### PR TITLE
Make BertClassifier operate directly on raw string inputs

### PR DESCRIPTION
**Pull request by mattdangerw**
This is now ready for review!

Demo based on this branch:
https://colab.sandbox.google.com/gist/mattdangerw/84f0818d902d666fe4c164f8108040c2/simplified-pipeline-demo.ipynb

### CODE REVIEW

File: keras_nlp/models/bert/bert_tasks_test.py [hunk 1]
```diff
@@ -34,82 +36,82 @@ def setUp(self):
...
+        self.raw_batch = tf.constant(
+            [
+                "the quick brown fox.",
+                "the slow brown fox.",
+                "the smelly brown fox.",
+                "the old brown fox.",
+            ]
+        )
+        self.preprocessed_batch = self.preprocessor(self.raw_batch)
+        self.raw_dataset = tf.data.Dataset.from_tensor_slices(
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_tasks_test.py:65]: Do we have enough testing (either here or in `pipeline_model_test.py`) for non-`Dataset` inputs? A user browsing this code could be forgiven for thinking `Dataset` was all we supported.

File: keras_nlp/models/bert/bert_tasks.py [hunk 2]
```diff
@@ -123,44 +204,70 @@ def from_preset(
     ):
         """Create a classification model from a preset architecture and weights.
 
+        By default, this method will automatically create a `preprocessor`
+        layer to preprocess raw inputs during `fit()`, `predict()`, and
+        `evaluate()`. If you would like to disable this behavior, pass
+        `preprocessor=None`.
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_tasks.py:210]: Since we can't use the same flag here as the base class (`include_preprocessing`), I appreciate the note!

Should we also include an example of specifying the preprocessor with overrides?
```python
preprocessor=BertPreprocessor.from_preset(
    "bert_base_uncased_en",
    sequence_length=128,
)
```

File: keras_nlp/models/bert/bert_tasks.py [hunk 3]
```diff
@@ -123,44 +204,70 @@ def from_preset(
...
 
         Examples:
+
+        Raw string inputs.
         ```python
-        input_data = {
+        # Create a dataset with raw string features in an `(x, y)` format.
+        features = ["The quick brown fox jumped.", "I forgot my homework."]
+        labels = [0, 3]
+        ds = tf.data.Dataset.from_tensor_slices((features, labels)).batch(2)
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_tasks.py:224]: Do we need `Dataset` conversions in the docstring? Seems kinda heavy

**Review by jbischof — COMMENTED** [review]
Very impressive, @mattdangerw! Some small comments to get things started.

File: keras_nlp/models/bert/bert_tasks_test.py (hunk 1 above)
Comment (jbischof) [reply keras_nlp/models/bert/bert_tasks_test.py:65]: My suspicion seems to be correct. Both the following calls throw in this branch:
```python
classifier = keras_nlp.models.BertClassifier.from_preset("bert_tiny_uncased_en")
classifier("I love modular workflows in keras-nlp!")
classifier(["I love modular workflows in keras-nlp!", "I hate spaghetti code."])
```

The only thing I could get to work outside `tf.data.Dataset` was a tensor input:
```python
classifier(tf.constant("I love modular workflows in keras-nlp!", shape=(1,))
```

Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks_test.py:65]: There is a good amount of testing for non-dataset inputs (especially for the pipeline utility itself, which is why that test file is fairly bulky). We could add more testing for raw inputs like this. But I think what you are hitting here is a few bugs that we need to clean up.
 - For the first call you show, I would expect that to hit our issue where our models will give a confusing error for unbatched inputs. That is https://github.com/keras-team/keras-nlp/issues/395
 - For the second call, that is an issue where that tuple is is interpreted as representing multiple segments, not multiple samples. So that call two is unbatched. That is https://github.com/keras-team/keras-nlp/issues/396
 - The third call you show is the only one that will result in a valid batch dimension, so the only one that does not error out here.

Clearly the UX is bad there. I can open some PRs with proposed fixes for both of those today. But I think the behavior you are describing is correct for the current state of things on master.

File: keras_nlp/models/bert/bert_tasks.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks.py:202]: done!

File: keras_nlp/models/bert/bert_tasks.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks.py]: Updated this to stop showing tf datasets, and just show passing in tensor/python inputs.

File: keras_nlp/models/bert/bert_tasks_test.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks_test.py:65]: Just FYI, today, if you wanted to pass a raw string. You would need to have a call like `classifier([["I love modular workflows in keras-nlp!"]])` (though I'm not advocating we keep that behavior!). First list is batch dim, second list is segment enumeration.

Comment (jbischof) [reply keras_nlp/models/bert/bert_tasks_test.py:65]: Thanks for the clarification @mattdangerw! At a minimum we shouldn't be throwing. Good communication with the user will be a whole other level of challenge.

Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks_test.py:65]: I would actually love for every usage you showed to "just work," as I think that would be the most intuitive experience there. But can follow up on PRs for those respective issues.

File: keras_nlp/models/bert/bert_tasks.py [hunk 4]
```diff
@@ -174,8 +291,11 @@ def from_preset(
                 f"""{", ".join(cls.presets)}. Received: {preset}."""
             )
 
+        if "preprocessor" not in kwargs:
+            kwargs["preprocessor"] = BertPreprocessor.from_preset(preset)
+
         # Check if preset is backbone-only model
-        if preset in cls.backbone_presets:
+        if preset in Bert.presets:
```
Comment (mattdangerw) [inline keras_nlp/models/bert/bert_tasks.py:298]: This felt cleaner to me than the `backbone_presets` added in the last PR. The `Bert` class should be the thing that enumerates what can be called for `Bert.from_preset`.

Comment (jbischof) [reply keras_nlp/models/bert/bert_tasks.py:298]: Sorry, already merged the PR! Perhaps one compromise could be 
```python
def backbone_presets(cls):
    return Bert.presets()
```

I think the property is needed otherwise it will be very hard to abstract out the `from_preset` logic.

File: keras_nlp/models/bert/bert_tasks.py [hunk 5]
```diff
@@ -90,35 +148,48 @@ def __init__(
...
             "name": self.name,
             "trainable": self.trainable,
         }
 
     @classmethod
     def from_config(cls, config):
-        if "backbone" in config:
+        if "backbone" in config and isinstance(config["backbone"], dict):
             config["backbone"] = keras.layers.deserialize(config["backbone"])
+        if "preprocessor" in config and isinstance(
```
Comment (mattdangerw) [inline keras_nlp/models/bert/bert_tasks.py:185]: There is an interesting consequence of rebasing over https://github.com/keras-team/keras-nlp/pull/494.

We can no longer either restore entirely from config or entirely form the main constructor. The backbone needs to be restored via `deserialize`, but the `preprocessor` cannot be restored via `deserialize`, because the vocab need to be downloaded as part of the config (and that is only available via `from_preset`).

I don't think we need to solve this perfectly on this PR, as it more has to with our internals for loading than any actual usage.

If we try to solve the vocabulary loading problem, we could have an interesting setting of config all the way down...

```
"config": {
    "backbone": {
        "class_name": "keras_nlp>Bert",
        "config": {
            "vocabulary_size": 30522,
            "hidden_dim": 128,
            "intermediate_dim": 512,
            "num_layers": 2,
            "num_heads": 2,
            "max_sequence_length": 512,
            "num_segments": 2,
            "dropout": 0.1,
        },
    },
    "preprocessor": {
        "class_name": "keras_nlp>BertPreprocessor",
        "config": {
            "truncator": "waterfall",
            "sequence_length": 128,
        },
    },
    "num_classes": 2,
},
```

This would be interesting in that it would give you to use non-default preprocessing argument during fine-tuning and replicate them here.

Anyway, I propose we just open an issue for this, and track this on a follow up!

File: keras_nlp/models/bert/bert_tasks.py (hunk 4 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks.py:298]: Talked a bit offline. We may want to factor out a common base class with `from_preset` logic. But even in that case, I think we could pull all this info from the backbone class.

I think we can remove this property for now (it makes it simpler to support the sst2 preset for tokenizer/preprocessor). And solve any issue related to common base as a follow up.

File: keras_nlp/models/bert/bert_presets_test.py [hunk 6]
```diff
@@ -77,16 +77,30 @@ def test_backbone_output(self, load_weights):
...
+            "bert_tiny_uncased_en",
+            load_weights=load_weights,
+        )
+        # We don't assert output values, as the head weights are random.
+        model.predict(input_data)
+
+    @parameterized.named_parameters(
+        ("load_weights", True), ("no_load_weights", False)
+    )
+    def test_classifier_output_without_preprocessing(self, load_weights):
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_presets_test.py:91]: Should we still have a test for `__call__` here? Or is that covered elsewhere?

File: keras_nlp/models/bert/bert_tasks.py [hunk 7]
```diff
@@ -31,32 +33,81 @@
...
 
 @keras.utils.register_keras_serializable(package="keras_nlp")
-class BertClassifier(keras.Model):
-    """BERT encoder model with a classification head.
+class BertClassifier(PipelineModel):
+    """An end-to-end BERT model for classification tasks
+
+    This model attaches a classification head to a `keras_nlp.model.Bert`
+    backbone, mapping from the backbone outputs to logit output suitable for
+    a classification task. For the easiest usage of this model with pre-trained
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_tasks.py:41]: Can probably drop "the easiest".

File: keras_nlp/models/bert/bert_tasks.py [hunk 8]
```diff
@@ -31,32 +33,81 @@
...
     Args:
         backbone: A `keras_nlp.models.Bert` instance.
         num_classes: int. Number of classes to predict.
+        preprocessor: A `keras_nlp.models.BertPreprocessor` or `None`. If
+            `None`, this model will not apply preprocessing, and inputs should
+            be preprocessed before calling the model.
 
     Examples:
+
+    Raw string inputs.
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_tasks.py:61]: Is this usage we want to encourage? I might nix this example

File: keras_nlp/models/bert/bert_tasks_test.py [hunk 9]
```diff
@@ -34,49 +36,79 @@ def setUp(self):
             max_sequence_length=128,
             name="encoder",
         )
-        self.classifier = BertClassifier(self.backbone, 4, name="classifier")
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_tasks_test.py:37]: Should we still have a test with dense inputs?

**Review by jbischof — COMMENTED** [review]
Great work! Left some questions

File: keras_nlp/models/bert/bert_tasks_test.py (hunk 9 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks_test.py:37]: We do I think.

Dense inputs to call -> test_valid_call_classifier
Dense inputs to predict -> test_bert_classifier_predict_no_preprocessing
Dense inptus to fit -> test_bert_classifier_fit_no_preprocessing

Something else to cover?

**Review by jbischof — APPROVED** [review]
Great work, thank you! We will need to open issues for adding this to other tasks models in the codebase.
